bpo-37953: Fix ForwardRef equality checks

plokmijnuhby opened the pull request:
Ideally if we stick a ForwardRef in a dictionary we would like to reliably be able to get it out again.

<!-- issue-number: [bpo-37953](https://bugs.python.org/issue37953) -->
https://bugs.python.org/issue37953
<!-- /issue-number -->

plokmijnuhby committed:
Update typing.py

plokmijnuhby committed:
Update typing.py

Maybe this will cut down on hash collisions

gvanrossum reviewed (commented):
This looks reasonable, but let's make sure @ilevkivskyi agrees.

Can you add a test (preferably one that shows what's not working without the fix)?

It's fairly unorthodox to have a PR without a bug in the tracker -- can you create a bug describing the problem and link this to it? Also please add a news item.

brandtbucher commented on Lib/typing.py:
```diff
@@ -524,11 +524,10 @@ def _evaluate(self, globalns, localns):
 
     def __hash__(self):
-        return hash((self.__forward_arg__, self.__forward_value__))
+        return hash((self.__forward_arg__,))
```
Is this tuple still needed here?

gvanrossum replied:
Good eye -- it is not.

plokmijnuhby committed:
Added unit test

ilevkivskyi commented on Lib/typing.py:
```diff
@@ -486,7 +486,7 @@ def open_helper(file: str, mode: MODE) -> str:
 
 
-class ForwardRef(_Final, _root=True):
+class ForwardRef(_Final, _Immutable, _root=True):
```
Why is this needed? This will preserve _identity_ (not just value) after copying.

plokmijnuhby replied:
Well, it's not strictly needed, but there's no reason that two references to the same thing should refer to different objects.

ilevkivskyi replied:
> Well, it's not strictly needed

Then don't do this. This can break things in unexpected ways. Also this "thing" has internal state that does change.

ilevkivskyi reviewed (commented):
Mostly looks good, just one question.

Please also add a `NEWS` item.

brandtbucher commented on Lib/typing.py:
```diff
@@ -524,11 +524,10 @@ def _evaluate(self, globalns, localns):
             return NotImplemented
-        return (self.__forward_arg__ == other.__forward_arg__ and
-                self.__forward_value__ == other.__forward_value__)
+        return self.__forward_arg__ == other.__forward_arg__
```
So it seems best to keep the check for forward value, contingent upon them both existing:

```suggestion
        if self.__forward_evaluated__ and other.__forward_evaluated__:
            return (self.__forward_arg__ == other.__forward_arg__ and
                    self.__forward_value__ == other.__forward_value__)
        return self.__forward_arg__ == other.__forward_arg__
```

I think that this, combined with the new `__hash__`, is probably the most useful and least surprising implementation.

brandtbucher commented on Lib/typing.py:
```diff
@@ -524,11 +524,10 @@ def _evaluate(self, globalns, localns):
 
     def __hash__(self):
-        return hash((self.__forward_arg__, self.__forward_value__))
+        return hash((self.__forward_arg__,))
```
```suggestion
        return hash(self.__forward_arg__)
```

plokmijnuhby committed:
Update Lib/typing.py

Remove the tuple in forwardref hash

Co-Authored-By: Brandt Bucher <brandtbucher@gmail.com>

plokmijnuhby committed:
Update Lib/typing.py

Alright, I don't really agree with this but I seem to be in the minority, so I'll change it.

Co-Authored-By: Brandt Bucher <brandtbucher@gmail.com>

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

ilevkivskyi commented on Misc/NEWS.d/next/Library/2019-09-06-17-40-34.bpo-37953.db5FQq.rst:
```diff
@@ -0,0 +1,3 @@
+Forward references in the typing module now have different hash and eq methods. References will now compare equal properly if only one of them has been evaluated and the other hasn't. Hash values are now consistent across the lifetime of a forward reference. 
```
Use `` `__hash__` `` and `` `__eq__` `` instead of just `hash` and `eq`. Also this news item is too long, couple short sentences (or even one) should be enough.

plokmijnuhby replied:
I changed it, is that better?

ilevkivskyi commented on Lib/test/test_typing.py:
```diff
@@ -2734,6 +2734,17 @@ def test_get_type_hints_ClassVar(self):
+        
+        gth(foo)
+        self.assertEqual(fr, typing.ForwardRef('int'))
+        self.assertEqual(hash(fr), hsh)
```
Please add few more tests, e.g. checking also equality and hashes between references that got evaluated to different things etc.

plokmijnuhby committed:
Update typing.py

plokmijnuhby committed:
Update 2019-09-06-17-40-34.bpo-37953.db5FQq.rst

plokmijnuhby committed:
Update test_typing.py

Add extra tests from pull request #15650.

hongweipeng commented on Lib/test/test_typing.py:
```diff
@@ -2355,6 +2414,64 @@ def foo(a: Tuple['T']):
+
+        r1 = namespace1()
+        r2 = namespace2()
+        assert r1 is not r2
```
It would be better to replace `assert r1 is not r2` with `self.assertIsNot(r1, r2)`.

aeros replied:
Agreed, bare ``assert`` expression shouldn't be used in the tests.

hongweipeng commented on Lib/test/test_typing.py:
```diff
@@ -2355,6 +2414,64 @@ def foo(a: Tuple['T']):
+        def namespace2():
+            a = typing.ForwardRef('A')
+            A = a
+            # class A: pass
```
Remove this line `# class A: pass`.

aeros commented on Misc/NEWS.d/next/Library/2019-09-06-17-40-34.bpo-37953.db5FQq.rst:
```diff
@@ -0,0 +1 @@
+Forward references in the typing module now have improved __hash__ and __eq__ methods.
```
Could we slightly modify this news entry to be more specific to the actual class the methods are being added to? Also fixes the formatting to use codeblocks for ``__hash__`` and ``__eq__`` as well as adding a Sphinx role to ``ForwardReferences`` and ``typing``. This allows readers to click on them to be linked to the corresponding documentation. 

```suggestion
In :mod:`typing`, improved the ``__hash__`` and ``__eq__`` methods for
:class:`ForwardReferences`.
```

Note that this change also requires this to be split into a second line, which I've included in the suggestion. This is because it goes above the 80 character width limit for reST.

plokmijnuhby committed:
Update test_typing.py

plokmijnuhby committed:
Update Misc/NEWS.d/next/Library/2019-09-06-17-40-34.bpo-37953.db5FQq.rst

Co-Authored-By: Kyle Stanley <aeros167@gmail.com>

ilevkivskyi reviewed (approved):
Thanks for updates! LGTM now.
